src/librustc/mir/interpret/mod.rs

```diff
@@ -161,25 +163,37 @@ impl ::rustc_serialize::UseSpecializedDecodable for AllocId {}
 #[derive(RustcDecodable, RustcEncodable)]
 enum AllocKind {
     Alloc,
+    AllocAtPos,
     Fn,
     Static,
 }
 
 pub fn specialized_encode_alloc_id<
     'a, 'tcx,
-    E: Encoder,
+    E: TyEncoder,
+    M,
 >(
     encoder: &mut E,
     tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    cache: M,
     alloc_id: AllocId,
-) -> Result<(), E::Error> {
+) -> Result<(), E::Error>
+where
+    M: Fn(&mut E) -> &mut FxHashMap<AllocId, usize>
+{
     let alloc_type: AllocType<'tcx, &'tcx Allocation> =
         tcx.alloc_map.lock().get(alloc_id).expect("no value for AllocId");
     match alloc_type {
         AllocType::Memory(alloc) => {
+            if let Some(alloc_pos) = cache(encoder).get(&alloc_id).cloned() {
+                AllocKind::AllocAtPos.encode(encoder)?;
+                return encoder.emit_usize(alloc_pos);
+            }
+            let pos = encoder.position();
             trace!("encoding {:?} with {:#?}", alloc_id, alloc);
             AllocKind::Alloc.encode(encoder)?;
             alloc.encode(encoder)?;
+            cache(encoder).insert_same(alloc_id, pos);
```

Comment:
`insert_same()` doesn't seem what we want here. If this could be reached in a racy way then `pos` would not necessarily be the same. It would also mean that two encoders would write to the same stream. Something like `assert!(cache(encoder).insert(alloc_id, pos).is_none())` seems more appropriate. Correct me if I'm wrong.